\documentclass[a4paper,10pt]{article}
%%%
\usepackage{draftwatermark} % DRAFT Watermark
% \SetWatermarkAngle{0} default is 45.0
\SetWatermarkScale{5}
\usepackage{listings} % For code blocks
\usepackage[usenames,dvipsnames,svgnames,table]{xcolor} % For colorized boxes and text
\usepackage[colorlinks=true,urlcolor=blue]{hyperref} % To make HTML links blue
\usepackage{courier} % for Courier font
\usepackage[a4paper]{geometry} % for a4paper margins
\usepackage{graphicx} % for the inclusion of screenshots
\usepackage{dirtree} % for directory listings
% Setup listings package for shell script and have comments presented in blue
% Text will be in Lavendar boxes with black border in Courier monospaced font
\lstset{ %
  backgroundcolor=\color{Lavender},   % choose the background color; you must add \usepackage{color} or \usepackage{xcolor}
  basicstyle=\footnotesize\ttfamily,        % the size of the fonts that are used for the code
  breakatwhitespace=false,         % sets if automatic breaks should only happen at whitespace
  breaklines=true,                 % sets automatic line breaking
  captionpos=t,                    % sets the caption-position to bottom
  commentstyle=\color{DarkBlue},    % comment style
  %deletekeywords={...},            % if you want to delete keywords from the given language
  %escapeinside={\%*}{*)},          % if you want to add LaTeX within your code
  %extendedchars=true,              % lets you use non-ASCII characters; for 8-bits encodings only, does not work with UTF-8
  frame=single,                    % adds a frame around the code
  keepspaces=true,                 % keeps spaces in text, useful for keeping indentation of code (possibly needs columns=flexible)
  %keywordstyle=\color{blue},       % keyword style
  language=sh,                 % the language of the code
  %morekeywords={*,...},            % if you want to add more keywords to the set
  %numbers=left,                    % where to put the line-numbers; possible values are (none, left, right)
  %numbersep=5pt,                   % how far the line-numbers are from the code
  %numberstyle=\tiny\color{mygray}, % the style that is used for the line-numbers
  rulecolor=\color{black},         % if not set, the frame-color may be changed on line-breaks within not-black text (e.g. comments (green here))
  showspaces=false,                % show spaces everywhere adding particular underscores; it overrides 'showstringspaces'
  showstringspaces=false,          % underline spaces within strings only
  showtabs=false,                  % show tabs within strings adding particular underscores
  stepnumber=5,                    % the step between two line-numbers. If it's 1, each line will be numbered
  %stringstyle=\color{mymauve},     % string literal style
  tabsize=4,                       % sets default tabsize to 2 spaces
  %title=\lstname                   % show the filename of files included with \lstinputlisting; also try caption instead of title
}
\parskip=1em
%%%

\title{Implementation of a standalone Galaxy instance for the creation of re-usable bioinformatics processing pipelines}
\author{Daniel J. Shea}
\date{August 16th, 2014}

\begin{document}
\maketitle
\pagebreak
\section{Introduction}
Using yeast (Schizosaccharomyces Pombe and Saccharomyces Cervisiae) as our model organism, our research is conducted into understanding the biological processes underlying the formation, function, and epigenetic inheritance of silent chromatin domains.  Three major areas of research are related to RNAi-mediated assembly of heterochromatin, SIR-mediated assembly of silent chromatin, and the regulation of recombination within the ribosomal DNA repeats.  In order to improve the efficiency of data analysis via re-usable bioinformatics pipelines and to encourage collaboration and software re-use within the lab, we have chosen to implement a local instance of the Galaxy Project's open source web-based platform.

\subsection{Heterochromatin's role in the silencing of genes}
The physical organization of DNA in eukaryotes plays a role in the levels of genetic expression by restricting the physical access of transcriptional machinery to various regions of an organism's genome.  The organization of DNA is structured around histones, octamers composed of highly conserved proteins, H2A, H2B, H3, and H4.  These proteins form heterodimers H2A/H2B and H3/H4, respectively, that then further form tetramers.  In turn, these tetramers then combine to form an octamer that is referred to as a histone.  DNA winds around these histones forming a nucleosome, with small segments of DNA, referred to as linker DNA, between them.  Structured around this basic organization of DNA are regions of tightly or loosely condensed chromatin, referred to as heterochromatin and euchromatin respectively.  Heterochromatin plays a functional role in gene silencing through the physical restriction of access to DNA by transcriptional machinery involved in transcribing DNA into messenger RNA (mRNA).

\subsection{Benefits derived from the implementation of Galaxy within the lab}
Galaxy is an open, web-based platform for the analysis of data intensive biological research.  It provides an interactive and readily accessible web interface to many of the commonly used bioinformatics tools available today.  Computational pipelines may be constructed from within the web interface and then shared between users.  Work flows are accessible via the work flow editor, which allows for the visualization and editing of processing pipelines.  These work flows may then be abstracted for application to multiple data sets, and the work flows themselves can be shared between Galaxy users through the Galaxy web interface.  This allows researchers within our environment to compose work flows that may then be re-used or re-purposed by other collaborators in the lab with minimal effort.

\section{Computational server architectural overview}
Our computational server, \texttt{\footnotesize{chromosome.med.harvard.edu}}, serves as the platform upon which we run our Galaxy instance.  It should be noted however, that a local galaxy instance can be run on hardware with relatively modest specifications.  Depending upon the data sets and types of analyses performed, you may be constrained by the available amount RAM on your server when dealing with particularly large data sets or application of computationally intensive analysis such as attempting to perform de novo assembly.  The table below details the hardware configuration for chromosome.

\subsection{Hardware configuration}
\renewcommand{\arraystretch}{1.5}
\begin{tabular}{|r|p{.8\linewidth}|}
\hline
\multicolumn{2}{|c|}{\textbf{chromosome.med.harvard.edu}}\\
\hline
\textbf{CPU:} & (2) Intel(R) Xeon(R) CPU  X5660  @ 2.80GHz, 6 Physical cores (12 Hyper-threaded)\\
\hline
\textbf{Memory:} & (18) 4GB DDR3-800 DIMMs for a total of 72GB\\
\hline
\textbf{Disk:} & (8) 1.817 TB SATA Drives configured as a RAID-6 array, providing 10.908TB of usable space\\
\hline
\textbf{OS:} & Ubuntu 12.04.4 LTS\\
\hline
\end{tabular}

\section{Galaxy architectural overview}
Galaxy provides default web server and local database storage for standalone instances.  Depending on the use case, it is advised to reconfigure the instance to make use of more robust data storage and/or web server solutions.  For our particular instance, we are currently utilizing a postgresql database as our back-end storage and have chosen to utilize the proxy capability of nginx to serve website content.  By default, sqlite internal database files are utilized for the storage of data and computational work flows.  This proved problematic as our anticipated data sets were determined to quickly tax a small flat file database system such as sqlite.  Therefore, we opted to create a postgresql database wherein to store all of our data.  Details on how to accomplish this are detailed below in the database subsection.  Details on configuration of nginx as a proxy are outlined in the nginx configuration subsection.

User authentication and role based authentication are handled by the Galaxy server itself.  Larger installations with access to a centralized user identification management system such as LDAP or kerberos may be able to take advantage of identity caching through configuration of login caching and forwarding via apache or nginx.  Attempts to tie our local instance to existing LDAP authentication mechanisms proved to break the existing administrative functionality of the server.  These issues were raised with the development team of the Galaxy Project and it is currently being assessed as a possible feature request by Galaxy development team.

Below, we outline the necessary steps undertaken to install and configure our Galaxy standalone instance using an nginx proxy-based web server and a postgresql back-end database for the storage of our data.

\subsection{Galaxy, postgresql and nginx installation}
To install Galaxy, we can retrieve the latest stable version from the mercurial repository.  In order to do this, we must have the mercurial package installed.  \textit{(\textbf{Note:} Since we are using Ubuntu server as our linux distribution, we will demonstrate how to do this using the Ubuntu package management tools.  Please refer to your operating system's package management documentation if you are using a different distribution of linux.)}

\begin{lstlisting}
sudo apt-get install mercurial
\end{lstlisting}

Once mercurial is installed we may then install the latest stable Galaxy from the mercurial repository.

\begin{lstlisting}
hg clone https://bitbucket.org/galaxy/galaxy-dist/
cd galaxy-dist
hg update stable
\end{lstlisting}

This is enough to configure a simple standalone Galaxy instance utilizing the internal sqlite database and python webserver.  You can test the installation by attempting to start Galaxy by running the following command from within the \texttt{\footnotesize{galaxy-dist}} directory.

\begin{lstlisting}
sh run.sh
\end{lstlisting}

The initial run of Galaxy will pull necessary python .egg files and start a local instance of the Galaxy server.  We can now shut down the test instance as we will next install postgresql and configure Galaxy to rely on the external database as our back-end data store for all of our research data.  Shutdown of Galaxy may be performed by issuing a \texttt{\footnotesize{Ctrl-c}} in the terminal where \texttt{\footnotesize{sh run.sh}} was instantiated.

To configure postgresql we will need to install the postgresql package.

\begin{lstlisting}
sudo apt-get install postgresql
\end{lstlisting}

Further details regarding the configuration of postgresql may be obtained in your operating system's system administration documentation, or by accessing postgresql documentation available \href{http://www.postgresql.org/}{online}.  Once the database software is installed, we must create a database owned by the user id that will run the Galaxy server.  In our case the user id is \texttt{\footnotesize{galaxy}}.  \textit{(\textbf{Note:} The name of the database user created in postgresql should match whatever username is chosen.)}  Since the Galaxy platform controls its own schema within the database, we need to ensure that our galaxy user has the appropriate permissions and ownership of the the database for table creation.  This is controlled through the postgresql configuration.  Instructions on how to assign roles and permissions to a postgresql user may be referenced in the postgresql documentation.

The nginx webserver may be installed by issuing the following command.

\begin{lstlisting}
sudo apt-get install nginx
\end{lstlisting}

For more administrative configuration details concerning nginx, please look at the \href{http://nginx.org/en/docs/}{nginx documentation}.  We will breifly outline the steps we have taken to set up nginx as a proxy for Galaxy, however this is only one possible configuration.  Further details regarding use of nginx or apache may be referenced in the \href{https://wiki.galaxyproject.org/Admin/Config/Performance/ProductionServer}{Galaxy documentation}.

\subsection{Integration of postgresql with Galaxy}
Once a database has been created and we have created both the relevant galaxy user in the operating system and in the database authentication systems, we can begin to configure Galaxy for use with an external database.
Galaxy stores its relevant configuration information in a file located at the top-level directory \texttt{\footnotesize{galaxy-dist}} within a file named \texttt{\footnotesize{universe\_wsgi.ini}}.

First we should ensure that the Galaxy server will be accessible from external interfaces if we are planning on setting up a web accessible standalone instance.  This can be done by making the following change to the host entry within \texttt{\footnotesize{universe\_wsgi.ini}}:

\begin{lstlisting}
# The address on which to listen.  By default, only listen to
# localhost (Galaxy will not be accessible over the network).
# Use '0.0.0.0' to listen on all available network interfaces.
# host = 127.0.0.1}
host = 0.0.0.0
\end{lstlisting}

The next line we will need to edit is the \texttt{\footnotesize{database\_connection}} line.  This will point Galaxy to our newly created postgresql database.

\begin{lstlisting}
# Connect to a local postgresql database where database galaxy has been configured for
# access by the galaxy user. - djs 2014-05-06
# The SQLAlchemy PostgreSQL dialect has been renamed from 'postgres' to 'postgresql'.
# The new URL format is postgresql[+driver]://<user>:<pass>@<host>/<dbname>
database_connection = postgresql://galaxy:PASSWORD_GOES_HERE@localhost/galaxy
\end{lstlisting}

We will also enable the optional server side cursors which will improve the performance on large queries.  This option is only available if we are using postgresql and was a consideration when determining which database to use as our storage model.

\begin{lstlisting}
# If large database query results are causing memory or response time issues in
# the Galaxy process, leave the result on the server instead.  This option is
# only available for PostgreSQL and is highly recommended.
#database_engine_option_server_side_cursors = False
database_engine_option_server_side_cursors = True
\end{lstlisting}

The next three entries into our configuration will direct Galaxy where to manage tools for the tool shed repositories and where dependencies are to be stored on the server.

\begin{lstlisting}
# Tool config files, defines what tools are available in Galaxy.
# Tools can be locally developed or installed from Galaxy tool sheds.
#tool_config_file = tool_conf.xml,shed_tool_conf.xml
tool_config_file = tool_conf.xml,shed_tool_conf.xml
\end{lstlisting}

\begin{lstlisting}
# Default path to the directory containing the tools defined in tool_conf.xml.
# Other tool config files must include the tool_path as an attribute in the <toolbox> tag.
#tool_path = tools
tool_path = tools
\end{lstlisting}

\begin{lstlisting}
# Path to the directory in which managed tool dependencies are placed.  To use
# the dependency system, see the documentation at:
# http://wiki.g2.bx.psu.edu/Admin/Config/Tool%20Dependencies
tool_dependency_dir = tool_dependencies
\end{lstlisting}

Next, we will need to configure Galaxy to take advantage of Data Libraries, a method for sharing data within the lab.  Directories in the configuration file listed here are relative paths, meaning, they are assumed to exist underneath our local \texttt{\footnotesize{galaxy-dist}} directory.  It should be noted that if you define a directory within the configuration file, you must still create the directory yourself and ensure it has appropriate permissions for access by Galaxy.  in our case the owner of the Galaxy processes is the \texttt{\footnotesize{galaxy}} user, so we created the relevant directories and ensured that the ownership was set appropriately.

Below is the entry in the configuration file telling Galaxy where files will be placed in order for the Galaxy administrator to upload them into Galaxy.

\begin{lstlisting}
# -- Data Libraries

# These library upload options are described in much more detail in the wiki:
# http://wiki.g2.bx.psu.edu/Admin/Data%20Libraries/Uploading%20Library%20Files

# Add an option to the library upload form which allows administrators to
# upload a directory of files.
library_import_dir = MoazedLabData/Processed
\end{lstlisting}

And here is the relevant system directory, showing the appropriate permissions for that directory.

\begin{lstlisting}
root@chromosome:/usr/local/galaxy-dist# ls -ld MoazedLabData/
drwxr-xr-x 4 galaxy galaxy 89 Jun 12 14:22 MoazedLabData/
\end{lstlisting}

We will now enable the functionality that allows us to paste system specific paths via the Data Library management tool.  Since our galaxy user is restricted to the Galaxy installation itself, this does not pose a security risk.  However, it is strongly advised that you carefully consider the implications of enabling this feature if you choose to run galaxy with a normal account.  It is advised that you create a galaxy account for running galaxy and restrict the galaxy user to only allow access to relevant system files.  Please refer to your operating system's system administration documentation if you are unsure how to do this, as it is outside the scope of this article.

\begin{lstlisting}
# Add an option to the admin library upload tool allowing admins to paste
# filesystem paths to files and directories in a box, and these paths will be
# added to a library.  Set to True to enable.  Please note the security
# implication that this will give Galaxy Admins access to anything your Galaxy
# user has access to.
allow_library_path_paste = True
\end{lstlisting}

Before we open up our Galaxy instance to users, it is very important that we change the default \texttt{\footnotesize{id\_secret}} in the configuration file.  Failing to do so poses a security risk.  This value should be some random string, in this example I have generated a randomized string of data for use as the secret using the apg utility.  If you do not have the apg utility installed on your system, in Ubuntu, it can be added by issuing the following command:

\begin{lstlisting}
sudo apt-get install apg
\end{lstlisting}

Now we will want to use apg to generate our \texttt{\footnotesize{id\_secret}}.  This can be done by issuing the following command:

\begin{lstlisting}
apg -a 1 -m 32 -x 32
\end{lstlisting}

\texttt{\footnotesize{-a}} tells \texttt{\footnotesize{apg}} to use random characters instead of attempting to provide a pronounceable password.  \texttt{\footnotesize{-m}} and \texttt{\footnotesize{-x}} are flags indicating the minimum and maximum lengths of the generated passwords.  Here we have chosen 32 for both values so that all passwords generated will be 32 characters in length.

Below is a sample output from the above referenced command.

\begin{lstlisting}
]d_@{MF+cgItXmc+s[{J64eAKQKB{/6U
~D!N$M(\oF%*q1I!WSrfEydZ6MaK.-.`
g~V?"A\=,T.,9hi:]4e]<$8~y`B\+uJJ
lA=ivi3>GF09OB9qNbeGaDhp#eaO(q%j
t=Y!l_Ip^NsT!46U+>tJ##^ug($H{en<
9K%hA=P]F^x[5M@n2tp$S(ir-akB-%yQ
\end{lstlisting}

Each line represents a randomly generated 32 character password.  Select one and place that string in the \texttt{\footnotesize{id\_secret}} entry within the \texttt{\footnotesize{universe\_wsgi.ini}} configuration file as shown below.  This key is used by Galaxy to generate internal values involved with cookies and URLs.  It must be changed or your Galaxy installation will not be secure from malicious users who could generate fake cookies using the default entry to gain access to a user's Galaxy account and thereby their data!

\begin{lstlisting}
# -- Users and Security

# Galaxy encodes various internal values when these values will be output in
# some format (for example, in a URL or cookie).  You should set a key to be
# used by the algorithm that encodes and decodes these values.  It can be any
# string.  If left unchanged, anyone could construct a cookie that would grant
# them access to others' sessions.
#id_secret = USING THE DEFAULT IS NOT SECURE!
id_secret = g~V?"A\=,T.,9hi:]4e]<$8~y`B\+uJJ
\end{lstlisting}

The next setting we may need to change is to normalize emails.  Galaxy user ids are constructed from the user's email address given at registration time.  Normalization will force the email address to all lowercase and avoid any potential case issues when a user enters their email address to login to the system.

\begin{lstlisting}
# If your proxy and/or authentication source does not normalize e-mail
# addresses or user names being passed to Galaxy - set the following option
# to True to force these to lower case.
normalize_remote_user_email = True
\end{lstlisting}

It is important if you are planning on administrating your galaxy instance to add your own email address to this entry and to create a user with a matching email address.  This will enable you to access the administrative functions later when bringing up the Galaxy instance and is important to be able to create the shared Data Libraries.
\begin{lstlisting}
# Administrative users - set this to a comma-separated list of valid Galaxy
# users (email addresses).  These users will have access to the Admin section
# of the server, and will have access to create users, groups, roles,
# libraries, and more.  For more information, see:
# http://wiki.g2.bx.psu.edu/Admin/Interface
#admin_users = None
admin_users = your.email@university.edu
\end{lstlisting}

Here we disable anonymous login, as only registered users will be given access to the laboratory instance of Galaxy.
\begin{lstlisting}
# Force everyone to log in (disable anonymous access).
require_login = True
\end{lstlisting}

Currently, we have left this as the default, allowing anyone to register an account with our instance.  Since our instance is behind a firewall an only accessible by individuals with access to the Private HMS network, we have decided to let our users create their own accounts and specify the email they would like to use.  Depending upon your network configuration and whether or not your Galaxy instance is accessible over the internet, you may wish to change this value to \texttt{False}.
\begin{lstlisting}
# Allow unregistered users to create new accounts (otherwise, they will have to
# be created by an admin).
#allow_user_creation = True
\end{lstlisting}

We have enabled administrative functionality to delete accounts.  By default it is not enabled.  Due to our rather lenient account creation policy we felt it necessary should individuals accidentally create a second account for themselves.
\begin{lstlisting}
# Allow administrators to delete accounts.
#allow_user_deletion = False
allow_user_deletion = True
\end{lstlisting}

\subsection{Integration of nginx for use as a proxy with Galaxy}
Nginx is a webserver software package that can be utilized as a proxy to serve static html and to provide improved server response in comparison to the built-in python webserver that Galaxy deploys natively.  Here, we will detail the necessary configuration steps that were required to enable Galaxy to work with nginx as a proxy.

The configuration file for the nginx webserver are located in \texttt{\footnotesize{/etc/nginx/nginx.conf}}.  This file will require the following changes in order for nginx to serve as a proxy for Galaxy.
Within the \texttt{\footnotesize{http}} clause, the following settings will allow nginx to utilize gzip compression to reduce the amount of data that needs to be passed through the proxy.  It is advisable to add this to the configuration if it is not already present as it significantly speeds up performance of the webserver when dealing with large data sets.

\begin{lstlisting}
        ##
        # Gzip Settings
        ##

        gzip on;
        gzip_disable "msie6";

        gzip_vary on;
        gzip_proxied any;
        gzip_comp_level 6;
        gzip_buffers 16 8k;
        gzip_http_version 1.1;
        gzip_types text/plain text/css application/json application/x-javascript text/xml application/xml application/xml+rss text/javascript;
\end{lstlisting}

These filters must also be enabled in the Galaxy configuration file \texttt{\footnotesize{universe\_wsgi.ini}} as well, as they are disabled by default.  Here, the proxy filters are defined near the top of the configuration file and we have enabled them.

\begin{lstlisting}
# ---- Filters --------------------------------------------------------------

# Filters sit between Galaxy and the HTTP server.

# These filters are disabled by default.  They can be enabled with
# 'filter-with' in the [app:main] section below.

# Define the gzip filter.
[filter:gzip]
use = egg:Paste#gzip

# Define the proxy-prefix filter.
[filter:proxy-prefix]
use = egg:PasteDeploy#prefix
prefix = /
\end{lstlisting}

Next, we comment out the entry for the default built-in webserver and enable filtering with a proxy.

\begin{lstlisting}
# If not running behind a proxy server, you may want to enable gzip compression
# to decrease the size of data transferred over the network.  If using a proxy
# server, please enable gzip compression there instead.
#filter-with = gzip

# If running behind a proxy server and Galaxy is served from a subdirectory,
# enable the proxy-prefix filter and set the prefix in the
# [filter:proxy-prefix] section above.
filter-with = proxy-prefix
\end{lstlisting}

Back within the \texttt{\footnotesize{/etc/nginx/nginx.conf}} file, we will now define the Galaxy application and direct the proxy to the location of static files to be served.
Since we are only running Galaxy on this server's web server, our configuration is rather minimal.  We serve pages on port 8080 (the default Galaxy port) and point the aliases to the appropriate directories underneath the Galaxy base installation directory \texttt{\footnotesize{/usr/local/galaxy-dist}}.

\begin{lstlisting}
        ##
        # Galaxy configuration
        ##
        upstream galaxy_app {
        server localhost:8080;
        }
        server {
                client_max_body_size 10G;
                # ... other server stuff ...
                location / {
                proxy_set_header X-Forwarded-Host $host;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                proxy_set_header X-URL-SCHEME https;
                }
                location /static {
                alias /usr/local/galaxy-dist/static;
                }
                location /static/style {
                alias /usr/local/galaxy-dist/static/june_2007_style/blue;
                }
                location /static/scripts {
                alias /usr/local/galaxy-dist/static/scripts/packed;
                }
                location /favicon.ico {
                alias /usr/local/galaxy-dist/static/favicon.ico;
                }
                location /robots.txt {
                alias /usr/local/galaxy-dist/static/robots.txt;
                }
        }
\end{lstlisting}

Once you have finished configuration of your Galaxy instance you can again bring it up and test that the configuration settings are working as intended.  You can verify that the Galaxy server is able to access your postgresql database and a new schema should appear within the database now that Galaxy is managing its internal storage via the postgresql database.  There is an important log associated with startup that should be mentioned at this point, as it will assist in troubleshooting any potential configuration issues and will allow you to determine when Galaxy has successfully started.

\section{Galaxy system service configuration}
Since Galaxy is running as a system service, we will make several additional operating system level modifications to our server.  The following subsections briefly outline such configuraton changes with respect to Ubuntu 12.04LTS.  Postgresql and nginx should already be enabled as system services that start at boot time as part of their installation via the package manager, therefore we will only cover the necessary steps required to ensure that Galaxy is also enabled and brought up as a system service.

\subsection{Galaxy log file rotation using logrotate}
Located under the galaxy installation (in our case \texttt{\footnotesize{/usr/local/galaxy-dist}})  we will find a log file named \texttt{\footnotesize{paster.log}}.  This file is where the Galaxy instance will log messages while it is running on the system.

On linux based systems, it might be beneficial to utilize \texttt{\footnotesize{logrotate}} to handle ensuring that \texttt{\footnotesize{paster.log}} is routinely rotated and compressed, to avoid continuously logging to a single file that will, over time, grow and consume system resources such as disk space.  This can be done by implementing a logrotate configuration for galaxy by placing a file named \texttt{\footnotesize{galaxy}} in \texttt{\footnotesize{/etc/logrotate.d}} with the following configuration.

\begin{lstlisting}
/usr/local/galaxy-dist/paster.log {
  weekly
  rotate 8
  copytruncate
  compress
  missingok
  notifempty
}
\end{lstlisting}

This tells the logrotate d$\ae$mon to rotate paster.log weekly, keeping the last 8 logs, copying the existing file and truncating the current log, and then compressing the copied logfile to reduce disk utilization storing the files.  It is ok if the file is missing, and do not perform the above actions if the current log file is empty.  For more information on how to configure the logrotate d$\ae$mon itself, please refer to your operating system administration documentation.

\subsection{Automating start up and shut down of the Galaxy service}
Since Galaxy will be treated as a system service on our server, we have opted to create a simple start up and shut down script on our Ubuntu server utilizing the Ubuntu upstart system.  this ensures that when our system is brought down for maintenance, or re-started, that our Galaxy instance will be brought down cleanly during the shut down process and brought back up during the system's boot sequence.  Below are the contents of our script for reference.

\lstinputlisting{galaxy.txt}

The LSB header information contained in this script is specific to our Ubuntu system's startup scripts.  It allows the \texttt{\footnotesize{update-rc.d}} script to configure the startup and shutdown of Galaxy for the appropriate runlevels during system initialization and shut down.

\section{Account creation and initial login}
Once Galaxy has been successfully installed and brought online, we can then create our user.  This section briefly covers the GUI interface, as most of the functionality is documented in the Galaxy user documentation.
At this point the Galaxy instance should be running according to the previously outlined configuration and accessible via a web browser.
By default, Galaxy makes use of port \texttt{\footnotesize{8080}}. Therefore, in our case, to access Galaxy we simply direct our browser to the following URL \href{http://chromosome.med.harvard.edu:8080}{http://chromosome.med.harvard.edu:8080}

From here, you will want to create the administrative account that matches the email address you specified in your configuration in the \texttt{\footnotesize{admin\_users}} section of \texttt{\footnotesize{universe\_wsgi.ini}}.\vspace{1em}
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/00_login}}\vspace{1em}

From here you be greeted with a registration page that will allow you to create your account.\vspace{1em}
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/01_account_creation}}\vspace{1em}

Once you have successfully logged in, you will be greeted with the standard Galaxy page.  Please note that you should have an \texttt{\footnotesize{Admin}} menu visible for your user.  (You may need to restart your Galaxy instance for this to take effect.)\vspace{1em}\\
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/02_main_page}}\vspace{1em}

\section{Setting up a Data Library in Galaxy}
Galaxy allows for the configuration of data libraries in order to share data files.  The benefit of a data library is that access control may be permissioned on a user or group basis, if necessary.  Furthermore, data stored in a data library is only stored once within the database.  When users import data from a data library, a pointer to the original data is passed.  This reduces disk utilization on the server and circumvents duplication of large data sets shared among multiple users.
We chose to implement a Data Library for the shared publication data that has been generated by researcher's within our lab.  The benefits of this are several-fold.  We have a central repository within the laboratory for all of our sequencing data.  New lab members can easily examine work that has been done to date and thereby reduce the time required for them to acclimate themselves to the work flow within the lab.  Researcher's unfamiliar or just starting out utilizing bioinformatics tools are presented with a relatively intuitive GUI interface that can assist them in performing analyses on our lab's server.  The centralized repository of sequence data maintains annotation of the relevant NCBI GEO and SRA accession identifiers.  In order to import the SRA data in the proper file formats applicable to Galaxy, the \texttt{\footnotesize{sra-toolkit}} was used to convert SRA formatted data to SAM file format and then uploaded to Galaxy via the Data Library management interface.
The directory hierarchy for SRA data is as follows:\vspace{1em}
\dirtree{%
.1 MoazedLabData.
.2 Processed.
.2 SRAs.
.3 SRP\dots.
.3 \dots.
}\vspace{1em}
\texttt{\footnotesize{MoazedLabData}} resides under the base \texttt{\footnotesize{galaxy-dist}}.  SRAs are downloaded from NCBI and then processed via \texttt{\footnotesize{sra-toolkit}} for conversion to a bzip2-compressed SAM format (\texttt{\footnotesize{SRRnnnnn.sam.bz2}}).
In order to automate the conversion of SRA data to SAM format, some helper scripts were written in \texttt{\footnotesize{bash}} and are included below.
\lstinputlisting[title=process.sh]{Scripts/process.sh}
\lstinputlisting[title=sra2sam.sh]{Scripts/sra2sam.sh}

Once files are downloaded from NCBI, \texttt{\footnotesize{process.sh}} is called from within the directory containing the \texttt{\footnotesize{sra}} data files, this script passes each file to \texttt{\footnotesize{sra2sam.sh}} which calls \texttt{\footnotesize{sam-dump}} (part of the sra-toolkit) to output a compressed SAM file to the \texttt{\footnotesize{Processed}} directory.

These files are then uploaded into Galaxy via the Data Library Management interface within the Admin menu of Galaxy.  Once the data has been uploaded into Galaxy, is it now stored within our postgresql database.

\section{Bioinformatics tools and integration with the Galaxy tool shed}
There are numerous bioinformatics software tools available for use on Linux-based computing platforms.  Galaxy leverages these by providing python wrappers that will present form based entry for the various parameters associated with each tool.  Fortunately, the majority of these software packages already have pre-defined wrappers that can be accessed through the Galaxy tool shed.  Administration of the tool shed allows us as the administrator to install these tools in Galaxy and make them available to users.  It should be noted however, that the software tools themselves must also be installed and available on the Galaxy user's \texttt{\footnotesize{\$PATH}} in order to be executed by Galaxy.  Each software package will have its own requirements regarding acceptable version number and individual configuration.  For alignment and NGS mapping tools like \texttt{\footnotesize{bwa}}, \texttt{\footnotesize{tophat}}, \texttt{\footnotesize{tophat2}}, and \texttt{\footnotesize{BLAST}}, the local indexes and libraries can be maintained on the local installation server where Galaxy is installed.  A good reference regarding how to configure local indexes is available in the Galaxy documentation regarding \href{https://wiki.galaxyproject.org/Admin/DataPreparation}{data preparation}.

For the purposes of our laboratory environment and research regarding our model ogranism, \textit{Schizosaccaromyces pombe}, there is no available UCSC database entry regarding our model organism.  Therefore, support for adding our model organism's reference genome into these built in indexes is not available.  As such, a reference genome in \texttt{\footnotesize{fasta}} format was installed into a shared Data Library for use in our computational work flows.  

\section{Galaxy work flow creation}
Work flows are created either by extracting a work flow from an existing analysis, or by building a work flow within the work flow editor.  Work flows allow for the creation of a generalized pipeline for the processing of biological data using the many tools installed in the Galaxy tool shed.  This approach utilizes a graphical flowchart detailing the inputs and outputs derived from each tool.  Individual tool's outputs may be connected to the inputs of subsequent tools, forming a data pipeline.  These work flows may then be saved and shared with other users of the Galaxy platform.

Work flows are modeled after data flow diagrams found in software engineering, whereby an input data set is transformed through a series of applied functions, resulting in output derived from the series of transformations applied to the data in the model.  This allows for an abstract representation of the analysis to be performed on a type of data input which will result in data output of a specific format.  Further complex analysis can then be derived by chaining relatively simple work flows to construct larger, more complex pipelines.  Presented below is an example, illustrating the work flow creation process, utilizing the work flow editor.

\section{Example work flow}
As an example workflow, we have chosen to utilize the experimental data found in the Nature protocol ``Differential gene expression analysis of RNA-seq experiments with TopHat and Cufflinks''.  For details on where to obtain the sample data for this protocol, please reference the original paper.  In this analysis RNA-seq data from \textit{Drosophila melanogaster} will be mapped to the reference genome using tophat.  The expressed genes and transcripts will then be assembled utilizing cufflinks and cuffmerge will then merge the assemblies into a single transcript annotation.  This data may then be analyzed through the use of CummeRbund to plot differentially expressed genes.

Sample data for the protocol was downloaded to the server running our Galaxy instance.  This data was then uploaded into our Galaxy instance through the Data Library management interface.\vspace{1em}\\
\setlength{\fboxsep}{0.5em}
\newsavebox\lstbox
\begin{lrbox}{\lstbox}
\begin{minipage}[c][][c]{.98\textwidth}
\textbf{A note about \texttt{\footnotesize{tophat}}}\\
In order to run the tophat portion of the protocol, we had to apply a manual patch to fix option passing to tophat2.  The details of this patch are outlined \href{https://groups.google.com/forum/#!topic/tuxedo-tools-users/vG5Rn0IgxoA}{here}.  It is possible that this patch has already been incorporated into a newer release of the tophat software, however as of the time of this writing, we were running the most recent release of tophat.
\begin{lstlisting}[xleftmargin=2em,xrightmargin=2em,linewidth=\textwidth]
root@chromosome:~# tophat --version
TopHat v2.0.11
\end{lstlisting}
\end{minipage}
\end{lrbox}
\fcolorbox{black}{NavajoWhite}{\usebox\lstbox}

While the procedure for performing a differential gene analysis closely follows the protocol outlined in the Nature paper, there are some additional steps that must be considered when performing the analysis within the Galaxy environment.  The first thing to take note of is that our fastq data will need to be run through the fastqsanger groomer to convert the quality scores.  Once this is done, we then may run \texttt{\footnotesize{tophat2}} to align the RNA-seq reads to the \textit{Drosophila melanogaster} reference genome.  We must also import files that we wish to use into our current history, so that they may be passed as arguments within the tool dialogs.  Therefore it is important to import both the annotation file \texttt{\footnotesize{genes.gtf}} and \texttt{\footnotesize{genome.fa}} into the history, along with the RNA-seq fastq data files to be aligned.

\subsection{Align the RNA-seq reads to a reference genome using tophat2}
Once the files have been imported from our Data Library, we first run the \texttt{\footnotesize{fastqsanger groomer}} on each of the RNA-seq files.  We then select the \texttt{\footnotesize{tophat2}} tool and select the appropriate options to perform our alignment.

Below is a series of reference screenshot detailing the options selected for running tophat2 against the first set of paired-end reads in our RNA-seq data set.\\
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/09_tophat2}}\vspace{1em}
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/10_tophat2}}\vspace{1em}
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/11_tophat2}}\vspace{1em}
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/12_tophat2}}\vspace{1em}
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/13_tophat2}}\vspace{1em}
Once the run has completed, the history sidebar will be updated.  provided the run was successful and contained no errors during execution, you should see something similar to this in your history.\vspace{1em}\\
\setlength\fboxsep{0pt}\fbox{\includegraphics[scale=0.5]{Screenshots/14_tophat2_output}}\vspace{1em}\\
Clicking on the eye icon of the \texttt{\footnotesize{align summary}} will display a summary of the alignment in the main window.\vspace{1em}\\
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/15_tophat2_results}}

\subsection{Assemble expressed genes and transcripts using cufflinks}
Now that the reads have been aligned to the reference genome, we can make use of the \texttt{\footnotesize{cufflinks}} tool to assemble to expressed genes and transcripts.  Cufflinks will estimate the abundances and test for differential expression and regulation.\vspace{1em}\\
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/16_cufflinks}}\vspace{1em}\\
\setlength\fboxsep{0.5em}
\newsavebox\topbox
\begin{lrbox}{\topbox}
\begin{minipage}[c][][c]{.98\textwidth}
\textbf{A note about server resources and Galaxy's queuing system}\vspace{1em}

As with \texttt{\footnotesize{tophat}}, \texttt{\footnotesize{cufflinks}} is a computationally demanding program.  It will take some time for the analysis to complete.  We can observe system utilization through the linux utility \texttt{\footnotesize{top}} to ensure the system is not being overtaxed before starting multiple analyses.\vspace{1em}

This screenshot shows example output during the run of our 6 \texttt{\footnotesize{cufflinks}} processes.  Galaxy has queued some of the runs until such a time that there are adequate system resources available to execute them.  This is one of the features and benefits to incorporating the Galaxy platform into our lab environment.  We now have a de facto queuing scheme for computational work within the lab!\vspace{1em}
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/17_cufflinks_system_utilization}}
\end{minipage}
\end{lrbox}
\fcolorbox{black}{NavajoWhite}{\usebox\topbox}

Once cufflinks has completed.  Galaxy will update the history to show the results of the finished runs.  We can examine the contents of the \texttt{\footnotesize{assembled transcripts}} by clicking on the eye icon in box.\vspace{1em}\\
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/18_cufflinks_results}}

We can now run \texttt{\footnotesize{cuffmerge}} to merge the assembled transcripts into a single merged transcriptome annotation.

Here we will make use of our reference sequence and annotation.  The relevant options and settings for \texttt{\footnotesize{cuffmerge}} have been noted in the screenshots below.\vspace{1em}\\
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/19_cuffmerge}}\vspace{1em}
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/20_cuffmerge}}\vspace{1em}

Once \texttt{\footnotesize{cuffmerge}} has completed its execution, we now should have a fully annotated transcriptome with proper gene associations from our reference annotated genome.  We can examine the contents of the results by clicking on the eye icon of the \texttt{\footnotesize{merged transcripts}}.\vspace{1em}\\
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/21_cuffmerge_results}}\vspace{1em}

\subsection{Identify differentially expressed genes and transcripts using cuffdiff}
With the finished assembly of our transcriptome, we may utilize \texttt{\footnotesize{cuffdiff}} to identify the differentially expressed genes and transcripts from our RNA-seq data.  The \texttt{\footnotesize{cuffdiff}} tool may be utilized if you wish to perform plot generation and analysis outside of Galaxy using R.\vspace{1em}\\
\textbf{Differential expression using \texttt{\footnotesize{cuffdiff}}}\vspace{1em}\\
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/22_cuffdiff}}\vspace{1em}
Here, we specify the bam files of our accepted hits from our 2 conditions, each consisting of 3 replicates.\vspace{1em}\\
\textbf{Condition 1 with 3 replicates}\\
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/23_cuffdiff_condition_1}}\vspace{1em}
\textbf{Condition 2 with 3 replicates}\\
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/24_cuffdiff_condition_2}}\vspace{1em}
\textbf{The default options for cuffdiff}\\
\setlength\fboxsep{0pt}\fbox{\includegraphics[width=\textwidth]{Screenshots/25_cuffdiff_options}}\vspace{1em}

\subsection{Analysis of differential expression with cummeRbund}
CummeRbund is an R library that can be accessed through the R interpreter.  Here we will explore the distribution of expression levels and plot the two conditions utilizing a scatterplot.  We will also create a volcano plot to visualize the differentially expressed genes from the data we have obtained from our \texttt{\footnotesize{cuffdiff}} analysis.

One important thing to note is that the \texttt{\footnotesize{cuffdiff}} files in our Galaxy history will be named using a different naming convention than what \texttt{\footnotesize{cummeRbund}} is expecting.  Therefore, when we export our \texttt{\footnotesize{cuffdiff}} files, they will need to be renamed to the appropriate file-names that \texttt{\footnotesize{cummeRbund}} attempts to load in order to generate our plots.  I have written a \texttt{\footnotesize{python}} script to handle the renaming of the data files to the expected format using the following translation table between the two naming conventions.\vspace{1em}\\
\renewcommand{\arraystretch}{1.5}
\begin{tabular}{|l|l|}
\hline
\multicolumn{2}{|c|}{\textbf{Galaxy to cummeRbund name translation}}\\
\hline
\textbf{\texttt{\footnotesize{Galaxy Name}}} & \textbf{\texttt{\footnotesize{cummeRbund Name}}}\\
\hline
\texttt{\footnotesize{transcript\_FPKM\_tracking}} & \texttt{\footnotesize{isoforms.fpkm\_tracking}}\\
\hline
\texttt{\footnotesize{transcript\_differential\_expression\_testing}} & \texttt{\footnotesize{isoform\_exp.diff}}\\
\hline
\texttt{\footnotesize{gene\_FPKM\_tracking}} & \texttt{\footnotesize{genes.fpkm\_tracking}}\\
\hline
\texttt{\footnotesize{gene\_differential\_expression\_testing}} & \texttt{\footnotesize{gene\_exp.diff}}\\
\hline
\texttt{\footnotesize{TSS\_groups\_FPKM\_tracking}} & \texttt{\footnotesize{tss\_groups.fpkm\_tracking}}\\
\hline
\texttt{\footnotesize{TSS\_groups\_differential\_expression\_testing}} & \texttt{\footnotesize{tss\_group\_exp.diff}}\\
\hline
\texttt{\footnotesize{CDS\_FPKM\_tracking}} & \texttt{\footnotesize{cds.fpkm\_tracking}}\\
\hline
\texttt{\footnotesize{CDS\_FPKM\_differential\_expression\_testing}} & \texttt{\footnotesize{cds\_exp.diff}}\\
\hline
\texttt{\footnotesize{CDS\_overloading\_differential\_expression\_testing}} & \texttt{\footnotesize{cds.diff}}\\
\hline
\texttt{\footnotesize{promoters\_differential\_expression\_testing}} & \texttt{\footnotesize{promoters.diff}}\\
\hline
\texttt{\footnotesize{splicing\_differential\_expression\_testing}} & \texttt{\footnotesize{splicing.diff}}\\
\hline
\end{tabular}

Listed below are the contents of the renaming script for reference.\\
\lstinputlisting[language=python]{Scripts/Galaxy2Cummerbund.py}

\section{Acknowledgments}
I would like to express my gratitude to Dr. Danesh Moazed, Investigator, Howard Hughes Medical Institute and Professor of Cell Biology, Harvard Medical School, for providing me the opportunity to participate in a research internship this summer.  It has been a very educational and exciting experience.  My time with the lab has provided me with a wealth of hands-on experience and knowledge.

I would also like to personally thank Dr. Keith Connolly, Postdoctoral Fellow of the Moazed Lab, Harvard Medical School, for his mentoring and collaborative efforts during my research internship.  Dr. Connolly took the time to avail me of his current research and instruct me in various laboratory techniques.  His patience and guidance were instrumental to the success of this project.

And lastly, I would like to acknowledge all of the researchers and staff of the Moazed Lab for their assistance and valuable input as end-users and testers during the integration of Galaxy platform with the lab's computational server.
\end{document}